container-search/src/main/java/com/yahoo/search/searchers/ValidateNearestNeighborSearcher.java

```diff
@@ -0,0 +1,127 @@
+import com.yahoo.prelude.query.NearestNeighborItem;
+import com.yahoo.prelude.query.ToolBox;
+import com.yahoo.search.Query;
+import com.yahoo.search.Result;
+import com.yahoo.search.Searcher;
+import com.yahoo.search.query.ranking.RankProperties;
+import com.yahoo.search.result.ErrorMessage;
+import com.yahoo.search.searchchain.Execution;
+import com.yahoo.tensor.Tensor;
+import com.yahoo.tensor.TensorType;
+import com.yahoo.vespa.config.search.AttributesConfig;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * Validates any NearestNeighborItem query items.
+ *
+ * @author arnej
+ */
+@Beta
+public class ValidateNearestNeighborSearcher extends Searcher {
+
+    private Map<String, TensorType> validAttributes = new HashMap<>();
+
+    public ValidateNearestNeighborSearcher(AttributesConfig attributesConfig) {
+        for (AttributesConfig.Attribute a : attributesConfig.attribute()) {
+            TensorType tt = null;
+            if (a.datatype() == AttributesConfig.Attribute.Datatype.TENSOR) {
+                tt = TensorType.fromSpec(a.tensortype());
+            }
+            validAttributes.put(a.name(), tt);
+        }
+    }
+
+    @Override
+    public Result search(Query query, Execution execution) {
+        Optional<ErrorMessage> e = validate(query);
+        return e.isEmpty() ? execution.search(query) : new Result(query, e.get());
+    }
+
+    private Optional<ErrorMessage> validate(Query query) {
+        NNVisitor visitor = new NNVisitor(query.getRanking().getProperties(), validAttributes);
+        ToolBox.visit(visitor, query.getModel().getQueryTree().getRoot());
+        return visitor.errorMessage;
+    }
+
+    private static class NNVisitor extends ToolBox.QueryVisitor {
+
+        public Optional<ErrorMessage> errorMessage = Optional.empty();
+
+        private RankProperties rankProperties;
+        private Map<String, TensorType> validAttributes;
+
+        public NNVisitor(RankProperties rankProperties, Map<String, TensorType> validAttributes) {
+            this.rankProperties = rankProperties;
+            this.validAttributes = validAttributes;
+        }
+
+        @Override
+        public boolean visit(Item item) {
+            if (item instanceof NearestNeighborItem) {
+                validate((NearestNeighborItem) item);
+            }
+            return true;
+        }
+
+        private void setError(String description) {
+            errorMessage = Optional.of(ErrorMessage.createIllegalQuery(description));
+        }
+
+        private void validate(NearestNeighborItem item) {
+            int target = item.getTargetNumHits();
+            if (target < 1) {
+                setError(item.toString() + " has invalid targetNumHits");
+                return;
+            }
+            String qprop = item.getQueryRankFeatureName();
+            List<Object> rankPropValList = rankProperties.asMap().get(qprop);
```

Comment:
I think you should look up in rankFeatures instead of rankProperties.

Reply:
my first version did lookup in rankFeatures but the tensor wasn't there anymore; the "prepare" method in Ranking had been called by the time the validation was run, so it was moved to rankProperties, and renamed.  This took some hours of hair-pulling to figure out.  We think it's best to make a new searcher that only calls this prepare() step and add explicit constraints - in a separate PR.

Reply:
Yes, that stuff is a flacking mess :-/
Might be better to wait with cleanup until we can remove the old query binary serialization?
You can add a constraint to be before the spot where this transformation happens.

Reply:
query.prepare() is only called from GroupingExecutor as far as i can see; that seems like a disaster waiting to happen...

Reply:
Yep.